siliconcompiler/tools/klayout/klayout_operations.py

```diff
@@ -0,0 +1,263 @@
+import pya
+
+import sys
+import os
+
+
+def read_layout(stream_file):
+    print(f"[INFO] Reading '{stream_file}'")
+    layout = pya.Layout()
+    layout.read(stream_file)
+
+    return layout
+
+
+def __with_timestamps(schema):
+    sc_step = schema.get('arg', 'step')
+    sc_index = schema.get('arg', 'index')
+
+    return schema.get('tool', 'klayout', 'task', 'operations', 'var', 'timestamps',
+                      step=sc_step, index=sc_index) == ['true']
+
+
+def __do_cell_swap(parent, old_cell_idx, new_cell, checked):
+    if (parent.cell_index() in checked):
+        return 0
+
+    checked.append(parent.cell_index())
+    replacements = 0
+    for inst in parent.each_inst():
+        if (inst.cell_index == old_cell_idx):
+            inst.cell = new_cell
+            replacements += 1
+        else:
+            replacements += __do_cell_swap(inst.cell, old_cell_idx, new_cell, checked)
+    return replacements
+
+
+def swap_cells(base_layout, oldcell, newcell):
+    top_cell = base_layout.top_cell()
+    old_cell = base_layout.cell(oldcell)
+    new_cell = base_layout.cell(newcell)
+
+    if (old_cell is None):
+        return base_layout
+    if (new_cell is None):
+        return base_layout
+
+    checked = []
+    replacements = __do_cell_swap(top_cell, old_cell.cell_index(), new_cell, checked)
+    print(f"[INFO] Swapping '{old_cell.name}' to '{new_cell.name}' in "
+          f"'{top_cell.name}': {replacements} occurrences")
+    base_layout.delete_cell(old_cell.cell_index())
+
+    return base_layout
+
+
+def add_outline(base_layout, layer):
+    top_cell = base_layout.top_cell()
+    bbox = top_cell.bbox()
+
+    layer_info = base_layout.get_info(layer)
+    print(f"[INFO] Adding outline to '{top_cell.name}' on layer '{layer_info.to_s()}'")
+
+    shapes = top_cell.shapes(layer)
+    shapes.insert(pya.Box(bbox))
+
+    return base_layout
+
+
+def add_layout(base_layout, layout):
+    top_cell = base_layout.top_cell()
+
+    other_layout_top = layout.top_cell()
+
+    print(f"[INFO] Adding layout from '{other_layout_top.name}' to '{top_cell.name}'")
+    new_cell = base_layout.create_cell(other_layout_top.name)
+    new_cell.copy_tree(other_layout_top)
+
+    cell_inst = pya.CellInstArray(new_cell.cell_index(), pya.Trans())
+    top_cell.insert(cell_inst)
+
+    return base_layout
+
+
+def add_layout_top_top(base_layout, new_top_cell_name):
```

Comment:
Name seems a bit funny, can't tell if it's a typo or if it's intentional :) As far as I can tell this is creating a new top-level cell above the base layout, so not sure if you meant `add_layout_to_top`.